src/dns/query_impl.hpp

```diff
@@ -99,55 +110,77 @@ class QueryImpl {
     ~QueryImpl() {}
 
     // Private to enforce usage through issue()
-    QueryImpl(std::string query, std::string address,
-              std::function<void(Response)> f, Logger *lp,
+    QueryImpl(QueryClass dns_class, QueryType dns_type, std::string name,
+              std::function<void(Error, Response)> f, Logger *lp,
               evdns_base *base, Libs *lev, SharedPointer<bool> cancd)
         : callback(f), libs(lev), cancelled(cancd), logger(lp) {
 
         assert(base != nullptr && lev != nullptr);
+        if (dns_class != QueryClass::IN) throw UnsupportedClassError();
+
+        // Allow PTR queries
+        if (dns_type == QueryType::PTR) {
+            std::string s;
+            if ((s = measurement_kit::unreverse_ipv4(name)) != "") {
+                dns_type = QueryType::REVERSE_A;
+                name = s;
+            } else if ((s = measurement_kit::unreverse_ipv6(name)) != "") {
+                dns_type = QueryType::REVERSE_AAAA;
+                name = s;
+            } else
+                throw InvalidNameForPTRError();
+        }
```

Comment:
Note to self: think of how this is done. In particular PTR lookups require you to reverse the IP address and I am not 100% sure it should be done automatically as in above.